### ISSUE TITLE
Fix S4158 FP: A list filled in a local function is not empty

### ISSUE THREAD

**Issue by MarkMuth** [SonarSource/sonar-dotnet#7350]
Add to a list using a local method and SonarLint claims that the collection is known to be empty [S4158](https://rules.sonarsource.com/csharp/RSPEC-4158).

Add this code to any method of your choice and you'll see the warning. Run it e.g. in Visual Studio's C# Interactive Window to get the output: `List contains something!`.

```
    var list = new List<string>();

    void doSomeHardWork()
    {
        list.Add("something");
    }

    doSomeHardWork();

    list.ForEach(x => Console
        .WriteLine($"List contains {x}!"));
```

* C#
* Visual Studio 2022
* .NET Framework 4.8 and .NET Standard 2.0
* Windows 11

**Comment by zsolt-kolbay-sonarsource** [comment]
Hi @MarkMuth. That's indeed a False Positive, one that we've been aware of, but couldn't fix due to limitations in our Symbolic Execution engine. Recently several improvements were done on that engine, which now enable us to start working on this issue.
If you wish to follow the progress [check out this issue](https://github.com/SonarSource/sonar-dotnet/issues/4478). I'm closing this one as duplicate.
